File: frontend/src/types/openclawConfig.ts
export type OpenClawResourceType =
  | 'channel'
  | 'skill'
  | 'session_template'
  | 'log_policy'
  | 'agent'
  | 'scheduled_task';

export type OpenClawConfigMode = 'none' | 'bundle' | 'manual';

export interface OpenClawConfigResourceSummary {
  id: number;
  resource_type: OpenClawResourceType;
  resource_key: string;
  name: string;
  enabled: boolean;
  version: number;
}

export interface OpenClawConfigResource extends OpenClawConfigResourceSummary {
  user_id: number;
  description?: string;
  tags: string[];
  content: unknown;
  created_at: string;
  updated_at: string;
}

export interface UpsertOpenClawConfigResourceRequest {
  resource_type: OpenClawResourceType;
  resource_key: string;
  name: string;
  description?: string;
  enabled: boolean;
  tags: string[];
  content: unknown;
}

export interface OpenClawConfigBundleItem {
  resource_id: number;
  sort_order: number;
  required: boolean;
  resource?: OpenClawConfigResourceSummary;
}

export interface OpenClawConfigBundle {
  id: number;
  user_id: number;
  name: string;
  description?: string;
  enabled: boolean;
  version: number;
  items: OpenClawConfigBundleItem[];
  created_at: string;
  updated_at: string;
}

export interface UpsertOpenClawConfigBundleRequest {
  name: string;
  description?: string;
  enabled: boolean;
  items: OpenClawConfigBundleItem[];
}

export interface OpenClawConfigPlan {
  mode: OpenClawConfigMode;
  bundle_id?: number;
  resource_ids?: number[];
}

export interface OpenClawConfigCompilePreview {
  mode: OpenClawConfigMode | 'archive';
  bundle?: OpenClawConfigBundle;
  selected_resources: OpenClawConfigResourceSummary[];
  resolved_resources: OpenClawConfigResourceSummary[];
  auto_included: OpenClawConfigResourceSummary[];
  warnings: string[];
  env_names: string[];
  payload_sizes: Record<string, number>;
  total_payload_bytes: number;
  manifest: unknown;
}

export interface OpenClawInjectionSnapshot {
  id: number;
  instance_id?: number;
  user_id: number;
  mode: OpenClawConfigMode;
  bundle_id?: number;
  selected_resource_ids: number[];
  resolved_resources: OpenClawConfigResourceSummary[];
  manifest: unknown;
  env_names: string[];
  payload_sizes: Record<string, number>;
  secret_name?: string;
  status: string;
  error_message?: string;
  created_at: string;
  updated_at: string;
  activated_at?: string;
}

export const OPENCLAW_RESOURCE_TYPES: Array<{ value: OpenClawResourceType; label: string }> = [
  { value: 'channel', label: 'Channel' },
  { value: 'skill', label: 'Skill' },
  { value: 'session_template', label: 'Session Template' },
  { value: 'log_policy', label: 'Log Policy' },
  { value: 'agent', label: 'Agent' },
  { value: 'scheduled_task', label: 'Scheduled Task' },
];

File: backend/internal/models/openclaw_config.go
package models

import "time"

// OpenClawConfigResource stores a reusable OpenClaw bootstrap resource.
type OpenClawConfigResource struct {
	ID           int       `db:"id,primarykey,autoincrement" json:"id"`
	UserID       int       `db:"user_id" json:"user_id"`
	ResourceType string    `db:"resource_type" json:"resource_type"`
	ResourceKey  string    `db:"resource_key" json:"resource_key"`
	Name         string    `db:"name" json:"name"`
	Description  *string   `db:"description" json:"description,omitempty"`
	Enabled      bool      `db:"enabled" json:"enabled"`
	Version      int       `db:"version" json:"version"`
	TagsJSON     string    `db:"tags_json" json:"-"`
	ContentJSON  string    `db:"content_json" json:"-"`
	CreatedAt    time.Time `db:"created_at" json:"created_at"`
	UpdatedAt    time.Time `db:"updated_at" json:"updated_at"`
}

func (r OpenClawConfigResource) TableName() string {
	return "openclaw_config_resources"
}

// OpenClawConfigBundle stores a reusable bundle of OpenClaw resources.
type OpenClawConfigBundle struct {
	ID          int       `db:"id,primarykey,autoincrement" json:"id"`
	UserID      int       `db:"user_id" json:"user_id"`
	Name        string    `db:"name" json:"name"`
	Description *string   `db:"description" json:"description,omitempty"`
	Enabled     bool      `db:"enabled" json:"enabled"`
	Version     int       `db:"version" json:"version"`
	CreatedAt   time.Time `db:"created_at" json:"created_at"`
	UpdatedAt   time.Time `db:"updated_at" json:"updated_at"`
}

func (b OpenClawConfigBundle) TableName() string {
	return "openclaw_config_bundles"
}

// OpenClawConfigBundleItem stores a resource membership inside a bundle.
type OpenClawConfigBundleItem struct {
	ID         int  `db:"id,primarykey,autoincrement" json:"id"`
	BundleID   int  `db:"bundle_id" json:"bundle_id"`
	ResourceID int  `db:"resource_id" json:"resource_id"`
	SortOrder  int  `db:"sort_order" json:"sort_order"`
	Required   bool `db:"required" json:"required"`
}

func (i OpenClawConfigBundleItem) TableName() string {
	return "openclaw_config_bundle_items"
}

// OpenClawInjectionSnapshot stores the rendered bootstrap payload used by an instance.
type OpenClawInjectionSnapshot struct {
	ID                      int        `db:"id,primarykey,autoincrement" json:"id"`
	InstanceID              *int       `db:"instance_id" json:"instance_id,omitempty"`
	UserID                  int        `db:"user_id" json:"user_id"`
	Mode                    string     `db:"mode" json:"mode"`
	BundleID                *int       `db:"bundle_id" json:"bundle_id,omitempty"`
	SelectedResourceIDsJSON string     `db:"selected_resource_ids_json" json:"-"`
	ResolvedResourcesJSON   string     `db:"resolved_resources_json" json:"-"`
	RenderedManifestJSON    string     `db:"rendered_manifest_json" json:"-"`
	RenderedEnvJSON         string     `db:"rendered_env_json" json:"-"`
	SecretName              *string    `db:"secret_name" json:"secret_name,omitempty"`
	Status                  string     `db:"status" json:"status"`
	ErrorMessage            *string    `db:"error_message" json:"error_message,omitempty"`
	CreatedAt               time.Time  `db:"created_at" json:"created_at"`
	UpdatedAt               time.Time  `db:"updated_at" json:"updated_at"`
	ActivatedAt             *time.Time `db:"activated_at" json:"activated_at,omitempty"`
}

func (s OpenClawInjectionSnapshot) TableName() string {
	return "openclaw_injection_snapshots"
}
